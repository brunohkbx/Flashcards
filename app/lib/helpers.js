import { Platform } from 'react-native';

export const waitFor = ms => {
  return new Promise(resolve => setTimeout(resolve, ms));
};

export const getOS = () => {
  return Platform.OS;
};
