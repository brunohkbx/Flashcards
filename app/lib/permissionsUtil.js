import { Permissions } from 'expo';

export const obtainNotificationPermission = async () => {
  const { status: existingStatus } = await Permissions.getAsync(
    Permissions.NOTIFICATIONS
  );
  let finalStatus = existingStatus;

  if (existingStatus !== 'granted') {
    const { status } = await Permissions.askAsync(Permissions.NOTIFICATIONS);

    finalStatus = status;
  }

  return finalStatus;
};
