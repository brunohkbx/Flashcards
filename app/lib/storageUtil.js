import { AsyncStorage } from 'react-native';
import { pickBy, assign } from 'lodash/object';

const DECK_STORAGE_KEY = 'flashcards:deck';

export const getDecks = () => {
  return AsyncStorage.getItem(DECK_STORAGE_KEY);
}

export const createDeck = deck => {
  return AsyncStorage.mergeItem(DECK_STORAGE_KEY, JSON.stringify(deck));
}

export const deleteDeck = id => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => pickBy(decks, deck => deck.id !== id))
    .then(decks => AsyncStorage.setItem(DECK_STORAGE_KEY, JSON.stringify(decks)));
}

export const editDeck = deck => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => assign(decks, deck))
    .then(decks => AsyncStorage.setItem(
      DECK_STORAGE_KEY, JSON.stringify(decks)
    ));
}
