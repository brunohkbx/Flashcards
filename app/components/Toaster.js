import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Snackbar } from 'react-native-paper';
import { waitFor } from '../lib/helpers';

class Toaster extends Component {
  HIDE_ANIMATION_DURATION = 150
  queue = []

  state = {
    visible: false,
    message: ''
  }

  showMessage = message => {
    this.queue.push(message);

    if (this.state.visible) {
      this.setState(
        { visible: false },
        this.processQueueAfterAnimation
      );
    }
    else
      this.processQueue();
  };

  processQueue = () => {
    const { queue } = this;

    if (queue.length > 0) {
      this.setState({
        message: queue.shift(),
        visible: true,
      });
    }
  };

  processQueueAfterAnimation = () => {
    return waitFor(this.HIDE_ANIMATION_DURATION).then(this.processQueue);
  };

  handleDismiss = () => {
    const { onDismissCallback } = this.props;

    this.setState({ visible: false }, onDismissCallback);
  }

  render() {
    const { visible, message } = this.state;

    return (
      <Snackbar
        visible={visible}
        onDismiss={this.handleDismiss}
      >
        {message}
      </Snackbar>
    );
  }
}

Toaster.propTypes = {
  onDismissCallback: PropTypes.func,
  icon: PropTypes.string
};

export default Toaster;

