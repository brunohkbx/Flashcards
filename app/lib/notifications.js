import { Notifications } from 'expo';
import * as storageUtil from '../lib/storageUtil';

export const clearLocalNotification = () => {
  return storageUtil.unscheduleNotification()
    .then(Notifications.cancelAllScheduledNotificationsAsync);
};

export const scheduleLocalNotification = () => {
  return storageUtil.isScheduledNotification()
    .then(JSON.parse)
    .then(data => {
      if (data === null) {
        Notifications.cancelAllScheduledNotificationsAsync();

        let tomorrow = new Date();
        tomorrow.setDate(tomorrow.getDate() + 1);
        tomorrow.setHours(18);
        tomorrow.setMinutes(30);

        Notifications.scheduleLocalNotificationAsync(
          {
            title: 'Haven\'t seen you in a while',
            body: 'ðŸ‘‹ Don\'t forget to take your quiz today!',
            ios: {
              sound: true
            },
            android: {
              sound: true,
              priority: 'high',
              sticky: false,
              vibrate: true
            }
          },
          {
            time: tomorrow,
            repeat: 'day',
          }
        );

        storageUtil.scheduleNotification();
      }
    });
};
