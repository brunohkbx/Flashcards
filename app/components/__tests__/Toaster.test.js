import React from 'react';
import { shallow } from 'enzyme';
import Toaster from '../Toaster';
import * as helpers from '../../lib/helpers';

describe('Toaster', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
    }, propOverrides)

    const wrapper = shallow(<Toaster {...defaultProps} />);
    const wrapperInstance = wrapper.instance();

    const dismissSnackbar = () => {
      wrapper.find('withTheme(Snackbar)').simulate('dismiss');
    };

    return { wrapper, wrapperInstance, dismissSnackbar }
  };

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  describe('showMessage', () => {
    it('pushes a new message to the queue', () => {
      const mockQueue = {push: jest.fn()};
      const {wrapperInstance} = setup();
      wrapperInstance.queue = mockQueue;

      wrapperInstance.showMessage('foo');

      expect(mockQueue.push).toHaveBeenCalledWith('foo');
    });

    describe('When visible', () => {
      it('hides the snackbar and calls processQueueAfterAnimation', () => {
        const {wrapperInstance} = setup();
        jest.spyOn(wrapperInstance, 'processQueue');
        jest.spyOn(wrapperInstance, 'setState');
        wrapperInstance.setState({ visible: true });

        wrapperInstance.showMessage('foo');

        expect(wrapperInstance.setState).toHaveBeenCalledWith(
          { visible: false },
          wrapperInstance.processQueueAfterAnimation
        );
      });
    });

    describe('When it is not visible', () => {
      it('calls processQueue', () => {
        const { wrapperInstance } = setup();
        jest.spyOn(wrapperInstance, 'processQueue');
        wrapperInstance.setState({ visible: false });

        wrapperInstance.showMessage('foo');

        expect(wrapperInstance.processQueue).toHaveBeenCalled();
      });
    });
  })

  describe('processQueue', () => {
    describe('When queue has length greather than 0', () => {
      it('displays the snackbar with the first item of the queue', () => {
        const { wrapperInstance } = setup();
        jest.spyOn(wrapperInstance, 'setState');
        wrapperInstance.queue = ['foo'];

        wrapperInstance.processQueue();

        expect(wrapperInstance.setState).toHaveBeenCalledWith(
          {
            message: 'foo',
            visible: true
          }
        )
      });
    });
  });

  describe('processQueueAfterAnimation', () => {
    it('calls processQueue after HIDE_ANIMATION_DURATION', () => {
      const { wrapperInstance } = setup();
      jest.spyOn(helpers, 'waitFor');
      jest.spyOn(wrapperInstance, 'processQueue');

      wrapperInstance.processQueueAfterAnimation().then(() => {
        expect(helpers.waitFor).toHaveBeenCalledWith(
          wrapperInstance.HIDE_ANIMATION_DURATION
        )
        expect(wrapperInstance.processQueue).toHaveBeenCalled();
      })
    });
  });

  describe('handleDismiss', () => {
    it('hides the snackbar and calls the given callback', () => {
      const { wrapperInstance } = setup();
      jest.spyOn(wrapperInstance, 'setState');

      wrapperInstance.handleDismiss();

      expect(wrapperInstance.setState).toHaveBeenCalledWith(
        { visible: false},
        wrapperInstance.props.onDismissCallback
      )
    });
  });
});
