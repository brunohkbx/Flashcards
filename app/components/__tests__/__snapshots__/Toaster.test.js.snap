// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster renders properly 1`] = `
<withTheme(Snackbar)
  duration={1500}
  onDismiss={[Function]}
  visible={false}
/>
`;
